### PR TITLE
GraphViz migration to ValueGraph

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

Migrate GraphViz to ValueGraph

I'd just ask someone to test this locally, please 🙏 

I can't get `tuist graph` to work on my machine as I always get:
image

I have no idea what is wrong (I keep getting even after I made fresh install two weeks ago, so maybe my machine is fundamentally and irreparably flawed), so if anyone has an idea how to fix this/debug this, that'd be great.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
Nice work @fortmarek - thanks for migrating this over!

> I can't get tuist graph to work on my machine as I always get:

I wonder if it's Big Sur or maybe an installation issue with `graphviz`? Worth trying removing and re-installing graphviz.

```sh
brew uninstall graphviz
brew install graphviz
```

If that still fails, you can try changing the output format to be a `.dot` file and examining the results (text based).

```sh
tuist graph --format dot
```

---

While testing I noticed a few discrepancies: 

When it comes to filtering targets _(by explicitly specifying their name)_, there's a mismatch between the released version of Tuist and this PR where transitive dependencies are not included:

```sh
cd fixtures/ios_app_with_transitive_framework
tuist graph AppTests
```

Released Version:
graph


This PR:
graph

`Framework2` isn't included.


Unrelated to those changes, I also noticed the `tuist graph --skip-external-dependencies` _(even the realised version)_ produces a slightly odd graph (unless I misunderstood what it's meant to do)

```sh
cd fixtures/ios_app_with_transitive_framework
tuist graph --skip-external-dependencies
```

graph

**Comment by bolismauro** [comment]
@fortmarek I had the very same issue when fixing the graph command. I feel it is a permission issue.
The way I checked the result, besides testing, is by outputting the textual description with `po` and then use an online service to render the graph. 
I haven't been able to understand the reason of the issue though (I'm not very familiar with mac development and permissions)

**Comment by fortmarek** [comment]
I have a question regarding `ValueGraphTraverser` - what's the reason for `directTargetDependencies` returning only local target dependencies (ie it does not return direct target dependencies that are part of a different project)? I believe that naming is misleading and it lead to the bug where `tuist graph AppTests` does not include `Framework1-iOS`. If there is a valid reason, I am for renaming `directTargetDependencies` to `directLocalTargetDependencies` and keeping method `directTargetDependencies` which would actually include *all* direct target dependencies - even if they are a part of a different project. cc @kwridan @pepibumur

**Comment by fortmarek** [comment]
thanks @bolismauro! 

It might be related to permissions but I am not really sure how to fix it/where to look.
I have tried reinstalling + running `brew install graphviz --HEAD` to install graphviz from source, neither of those helped, unfortunately. So far I have resorted to opening `.dot` file with a VS Code extension to visualize the graph. But this might be worth investigating more as it seems this is not an issue applicable only to my setup but might be more widespread than we have previously thought.

**Comment by kwridan** [comment]
>  what's the reason for directTargetDependencies returning only local target dependencies (ie it does not return direct target dependencies that are part of a different project)

I believe there are some cases like searching for host apps (for tests or extensions) where only local targets need to be searched. 

> I am for renaming `directTargetDependencies` to `directLocalTargetDependencies` and keeping method directTargetDependencies which would actually include all direct target dependencies

+1 this makes sense to clarify the intent

**Comment by fortmarek** [comment]
all of the troubling scenarios you have listed should be fixed. I tested that locally leveraging `dot` format + added appropriate unit tests to catch any possible future regressions. I think this is worth another look @kwridan, thanks for the thorough review 🙇

File: Sources/TuistCore/GraphTraverser/GraphTraversing.swift [hunk 1]
```diff
@@ -62,12 +62,20 @@ public protocol GraphTraversing {
     /// - Parameter path: Path to the directory that contains the definition of the project.
     func targets(at path: AbsolutePath) -> Set<ValueGraphTarget>
 
-    /// Given a project directory and target name, it returns all its direct target dependencies.
+    /// Given a project directory and target name, it returns **all**l its direct target dependencies present in the same project.
+    /// If you want only direct target dependencies present in the same project as the target, use `directLocalTargetDependencies` instead
     /// - Parameters:
-    ///   - path: Path to the directory that contains the project.
+    ///   - path: Path to the directory that contains the target's project.
```
Comment (kwridan) [inline Sources/TuistCore/GraphTraverser/GraphTraversing.swift:68]: 👍

File: Tests/TuistCoreTests/ValueGraph/ValueGraphTraverserTests.swift [hunk 2]
```diff
@@ -9,6 +9,98 @@ import XCTest
 @testable import TuistSupportTesting
 
 final class ValueGraphTraverserTests: TuistUnitTestCase {
+    func test_dependsOnXCTest_when_is_framework() {
```
Comment (kwridan) [inline Tests/TuistCoreTests/ValueGraph/ValueGraphTraverserTests.swift:12]: nicely done

**Review by kwridan — APPROVED** [review]
awesome work @fortmarek

**Pull request merged**